File: apps/java/design-pattern/src/test/java/pattern/command/CommandTest.java
package pattern.command;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;

class CommandTest {

    @TempDir
    Path tempDir;

    @Test
    void createFileCommandCreatesFile() throws IOException {
        Path file = tempDir.resolve("test.txt");
        Command cmd = new CreateFileCommand(file, "hello");

        cmd.execute();

        assertTrue(Files.exists(file));
        assertEquals("hello", Files.readString(file));
        assertEquals("Create file: " + file, cmd.getDescription());
    }

    @Test
    void createFileCommandUnexecuteDeletesFile() throws IOException {
        Path file = tempDir.resolve("test.txt");
        Command cmd = new CreateFileCommand(file, "hello");

        cmd.execute();
        assertTrue(Files.exists(file));

        cmd.unexecute();
        assertFalse(Files.exists(file));
    }

    @Test
    void deleteFileCommandDeletesAndRestoresFile() throws IOException {
        Path file = tempDir.resolve("existing.txt");
        Files.writeString(file, "original content");

        Command cmd = new DeleteFileCommand(file);

        cmd.execute();
        assertFalse(Files.exists(file));

        cmd.unexecute();
        assertTrue(Files.exists(file));
        assertEquals("original content", Files.readString(file));
    }

    @Test
    void compositeCommandExecutesAndUnexecutesInOrder() throws IOException {
        Path file1 = tempDir.resolve("file1.txt");
        Path file2 = tempDir.resolve("file2.txt");

        CompositeCommand composite = new CompositeCommand();
        composite.addCommand(new CreateFileCommand(file1, "content1"));
        composite.addCommand(new CreateFileCommand(file2, "content2"));

        composite.execute();
        assertTrue(Files.exists(file1));
        assertTrue(Files.exists(file2));

        String desc = composite.getDescription();
        assertTrue(desc.contains("Create file: " + file1));
        assertTrue(desc.contains("Create file: " + file2));

        composite.unexecute();
        assertFalse(Files.exists(file1));
        assertFalse(Files.exists(file2));
    }

    @Test
    void slickButtonExecutesLambdaCommand() {
        AtomicReference<String> result = new AtomicReference<>("");
        SlickButton button = new SlickButton(() -> result.set("clicked"));

        button.onButtonPush();

        assertEquals("clicked", result.get());
    }

    @Test
    void slickButtonHandlesNullCommand() {
        SlickButton button = new SlickButton(null);
        assertDoesNotThrow(button::onButtonPush);
    }
}

File: apps/typescript/design-pattern/tests/command.test.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import {
  CreateFileCommand,
  DeleteFileCommand,
  CompositeCommand,
} from '../src/command';

describe('Command パターン', () => {
  let tmpDir: string;

  beforeEach(() => {
    tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'cmd-test-'));
  });

  afterEach(() => {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  });

  it('CreateFileCommand はファイルを作成する', () => {
    const filePath = path.join(tmpDir, 'test.txt');
    const cmd = new CreateFileCommand(filePath, 'hello');
    cmd.execute();

    expect(fs.existsSync(filePath)).toBe(true);
    expect(fs.readFileSync(filePath, 'utf-8')).toBe('hello');
  });

  it('CreateFileCommand の undo はファイルを削除する', () => {
    const filePath = path.join(tmpDir, 'test.txt');
    const cmd = new CreateFileCommand(filePath, 'hello');
    cmd.execute();
    cmd.undo();

    expect(fs.existsSync(filePath)).toBe(false);
  });

  it('DeleteFileCommand はファイルを削除する', () => {
    const filePath = path.join(tmpDir, 'test.txt');
    fs.writeFileSync(filePath, 'content');
    const cmd = new DeleteFileCommand(filePath);
    cmd.execute();

    expect(fs.existsSync(filePath)).toBe(false);
  });

  it('DeleteFileCommand の undo はファイルを復元する', () => {
    const filePath = path.join(tmpDir, 'test.txt');
    fs.writeFileSync(filePath, 'content');
    const cmd = new DeleteFileCommand(filePath);
    cmd.execute();
    cmd.undo();

    expect(fs.existsSync(filePath)).toBe(true);
    expect(fs.readFileSync(filePath, 'utf-8')).toBe('content');
  });

  it('CompositeCommand は全コマンドを順番に実行する', () => {
    const file1 = path.join(tmpDir, 'a.txt');
    const file2 = path.join(tmpDir, 'b.txt');
    const composite = new CompositeCommand([
      new CreateFileCommand(file1, 'aaa'),
      new CreateFileCommand(file2, 'bbb'),
    ]);
    composite.execute();

    expect(fs.existsSync(file1)).toBe(true);
    expect(fs.existsSync(file2)).toBe(true);
  });

  it('CompositeCommand の undo は逆順に取り消す', () => {
    const file1 = path.join(tmpDir, 'a.txt');
    const file2 = path.join(tmpDir, 'b.txt');
    const composite = new CompositeCommand([
      new CreateFileCommand(file1, 'aaa'),
      new CreateFileCommand(file2, 'bbb'),
    ]);
    composite.execute();
    composite.undo();

    expect(fs.existsSync(file1)).toBe(false);
    expect(fs.existsSync(file2)).toBe(false);
  });

  it('Command は description を持つ', () => {
    const cmd = new CreateFileCommand('/tmp/x.txt', '');
    expect(cmd.description).toBe('Create file: x.txt');
  });
});
